File: ios/PttTalk/Sources/PttTalkLib/EncryptedCallKeyCodec.swift
import Foundation

public enum EncryptedCallKeyMessageKind: UInt8, Sendable {
    case announcement = 1
    case acknowledgement = 2
}

public struct EncryptedCallKeyMessage: Equatable, Identifiable, Sendable {
    public let messageId: UUID
    public let channelId: UUID
    public let membershipEpoch: Int32
    public let callId: UUID
    public let callEpoch: Int32
    public let kind: EncryptedCallKeyMessageKind
    public let participantIdentity: String
    public let key: Data?
    public let senderAci: String
    public let senderDeviceId: Int
    public var id: UUID { messageId }

    public init(
        messageId: UUID = UUID(), channelId: UUID, membershipEpoch: Int32,
        callId: UUID, callEpoch: Int32, kind: EncryptedCallKeyMessageKind,
        participantIdentity: String, key: Data?, senderAci: String = "", senderDeviceId: Int = 0
    ) {
        self.messageId = messageId
        self.channelId = channelId
        self.membershipEpoch = membershipEpoch
        self.callId = callId
        self.callEpoch = callEpoch
        self.kind = kind
        self.participantIdentity = participantIdentity
        self.key = key
        self.senderAci = senderAci
        self.senderDeviceId = senderDeviceId
    }
}

public enum EncryptedCallKeyCodec {
    private static let magic = Data("PTTC".utf8)
    private static let version: UInt8 = 1
    private static let zeroUuid = UUID(uuid: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0))

    public static func encode(_ message: EncryptedCallKeyMessage) throws -> Data {
        guard message.messageId != zeroUuid, message.channelId != zeroUuid, message.callId != zeroUuid,
              message.membershipEpoch > 0, message.callEpoch > 0,
              let identity = message.participantIdentity.data(using: .utf8),
              (16...128).contains(identity.count),
              (message.kind == .announcement ? message.key?.count == 32 : message.key?.count == 16) else {
            throw EncryptedCallMediaError.invalidKey
        }
        var output = magic
        output.append(version)
        output.append(message.kind.rawValue)
        output.append(contentsOf: uuidBytes(message.messageId))
        output.append(contentsOf: uuidBytes(message.channelId))
        output.append(contentsOf: uuidBytes(message.callId))
        append(message.membershipEpoch, to: &output)
        append(message.callEpoch, to: &output)
        output.append(UInt8(identity.count))
        output.append(identity)
        output.append(message.key ?? Data())
        return output
    }

    public static func decode(_ bytes: Data, senderAci: String, senderDeviceId: Int) throws -> EncryptedCallKeyMessage {
        guard bytes.count >= 63, bytes.prefix(4) == magic, bytes[4] == version,
              let kind = EncryptedCallKeyMessageKind(rawValue: bytes[5]),
              UUID(uuidString: senderAci) != nil, (1...2).contains(senderDeviceId) else {
            throw EncryptedCallMediaError.invalidKey
        }
        var offset = 6
        let messageId = try readUuid(bytes, &offset)
        let channelId = try readUuid(bytes, &offset)
        let callId = try readUuid(bytes, &offset)
        let membershipEpoch = try readInt32(bytes, &offset)
        let callEpoch = try readInt32(bytes, &offset)
        guard offset < bytes.count else { throw EncryptedCallMediaError.invalidKey }
        let identityCount = Int(bytes[offset]); offset += 1
        guard (16...128).contains(identityCount), offset + identityCount <= bytes.count,
              let identity = String(data: bytes[offset..<offset + identityCount], encoding: .utf8) else {
            throw EncryptedCallMediaError.invalidKey
        }
        offset += identityCount
        let key = Data(bytes[offset...])
        guard membershipEpoch > 0, callEpoch > 0,
              (kind == .announcement ? key.count == 32 : key.count == 16) else {
            throw EncryptedCallMediaError.invalidKey
        }
        let decoded = EncryptedCallKeyMessage(
            messageId: messageId, channelId: channelId, membershipEpoch: membershipEpoch,
            callId: callId, callEpoch: callEpoch, kind: kind,
            participantIdentity: identity, key: key,
            senderAci: senderAci.lowercased(), senderDeviceId: senderDeviceId
        )
        guard try encode(decoded) == bytes else { throw EncryptedCallMediaError.invalidKey }
        return decoded
    }

    private static func uuidBytes(_ value: UUID) -> [UInt8] {
        withUnsafeBytes(of: value.uuid) { Array($0) }
    }

    private static func append(_ value: Int32, to output: inout Data) {
        var big = value.bigEndian
        withUnsafeBytes(of: &big) { output.append(contentsOf: $0) }
    }

    private static func readUuid(_ bytes: Data, _ offset: inout Int) throws -> UUID {
        guard offset + 16 <= bytes.count else { throw EncryptedCallMediaError.invalidKey }
        let raw = Array(bytes[offset..<offset + 16]); offset += 16
        return UUID(uuid: (
            raw[0], raw[1], raw[2], raw[3], raw[4], raw[5], raw[6], raw[7],
            raw[8], raw[9], raw[10], raw[11], raw[12], raw[13], raw[14], raw[15]
        ))
    }

    private static func readInt32(_ bytes: Data, _ offset: inout Int) throws -> Int32 {
        guard offset + 4 <= bytes.count else { throw EncryptedCallMediaError.invalidKey }
        let value = bytes[offset..<offset + 4].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
        offset += 4
        return Int32(bitPattern: value)
    }
}

/// Keychain-backed local inbox format used to hand call coordination safely across process restarts.
enum EncryptedCallKeyQueueCodec {
    private static let magic = Data("PTTQ".utf8)
    private static let version: UInt8 = 1
    static let maximumMessages = 128

    static func encode(_ messages: [EncryptedCallKeyMessage]) throws -> Data {
        guard messages.count <= maximumMessages else { throw EncryptedCallMediaError.invalidKey }
        let rows = try messages.map { message -> (EncryptedCallKeyMessage, Data, UUID) in
            guard let sender = UUID(uuidString: message.senderAci), (1...2).contains(message.senderDeviceId)
            else { throw EncryptedCallMediaError.invalidKey }
            return (message, try EncryptedCallKeyCodec.encode(message), sender)
        }
        var output = magic
        output.append(version)
        append(UInt16(rows.count), to: &output)
        for (message, body, sender) in rows {
            guard body.count <= Int(UInt16.max) else { throw EncryptedCallMediaError.invalidKey }
            withUnsafeBytes(of: sender.uuid) { output.append(contentsOf: $0) }
            output.append(UInt8(message.senderDeviceId))
            append(UInt16(body.count), to: &output)
            output.append(body)
        }
        return output
    }

    static func decode(_ bytes: Data?) throws -> [EncryptedCallKeyMessage] {
        guard let bytes, !bytes.isEmpty else { return [] }
        guard bytes.count >= 7, bytes.prefix(4) == magic, bytes[4] == version else {
            throw EncryptedCallMediaError.invalidKey
        }
        var offset = 5
        let count = Int(try readUInt16(bytes, &offset))
        guard count <= maximumMessages else { throw EncryptedCallMediaError.invalidKey }
        var messages: [EncryptedCallKeyMessage] = []
        messages.reserveCapacity(count)
        for _ in 0..<count {
            guard offset + 19 <= bytes.count else { throw EncryptedCallMediaError.invalidKey }
            let raw = Array(bytes[offset..<offset + 16]); offset += 16
            let sender = UUID(uuid: (
                raw[0], raw[1], raw[2], raw[3], raw[4], raw[5], raw[6], raw[7],
                raw[8], raw[9], raw[10], raw[11], raw[12], raw[13], raw[14], raw[15]
            )).uuidString.lowercased()
            let deviceId = Int(bytes[offset]); offset += 1
            let size = Int(try readUInt16(bytes, &offset))
            guard (1...2).contains(deviceId), (63...255).contains(size), offset + size <= bytes.count
            else { throw EncryptedCallMediaError.invalidKey }
            let body = Data(bytes[offset..<offset + size]); offset += size
            messages.append(try EncryptedCallKeyCodec.decode(
                body, senderAci: sender, senderDeviceId: deviceId
            ))
        }
        guard offset == bytes.count else { throw EncryptedCallMediaError.invalidKey }
        return messages
    }

    private static func append(_ value: UInt16, to output: inout Data) {
        var big = value.bigEndian
        withUnsafeBytes(of: &big) { output.append(contentsOf: $0) }
    }

    private static func readUInt16(_ bytes: Data, _ offset: inout Int) throws -> UInt16 {
        guard offset + 2 <= bytes.count else { throw EncryptedCallMediaError.invalidKey }
        let value = (UInt16(bytes[offset]) << 8) | UInt16(bytes[offset + 1])
        offset += 2
        return value
    }
}

File: android/talk/src/main/kotlin/app/ptt/talk/EncryptedCallKeyCodec.kt
package app.ptt.talk

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.UUID

private fun ByteArray?.contentEqualsNullable(other: ByteArray?): Boolean =
    if (this == null || other == null) this == null && other == null else contentEquals(other)

internal enum class EncryptedCallKeyMessageKind(val wire: Byte) {
    ANNOUNCEMENT(1), ACKNOWLEDGEMENT(2);

    companion object { fun fromWire(value: Byte) = entries.firstOrNull { it.wire == value } }
}

internal data class EncryptedCallKeyMessage(
    val messageId: UUID = UUID.randomUUID(),
    val channelId: UUID,
    val membershipEpoch: Int,
    val callId: UUID,
    val callEpoch: Int,
    val kind: EncryptedCallKeyMessageKind,
    val participantIdentity: String,
    val key: ByteArray?,
    val senderAci: String = "",
    val senderDeviceId: Int = 0,
) {
    override fun equals(other: Any?): Boolean = other is EncryptedCallKeyMessage &&
        messageId == other.messageId && channelId == other.channelId &&
        membershipEpoch == other.membershipEpoch && callId == other.callId &&
        callEpoch == other.callEpoch && kind == other.kind &&
        participantIdentity == other.participantIdentity && key.contentEqualsNullable(other.key) &&
        senderAci == other.senderAci && senderDeviceId == other.senderDeviceId

    override fun hashCode(): Int = 31 * messageId.hashCode() + (key?.contentHashCode() ?: 0)
}

internal enum class CallKeyMessageDisposition { PROCESS, DEFER_FUTURE_EPOCH, DISCARD }

/**
 * Roster updates and encrypted key envelopes use separate authenticated transports. A valid
 * next-epoch envelope can therefore arrive before the next authoritative call snapshot. Retain
 * it without installing it; stale and context-mismatched envelopes remain terminal.
 */
internal object CallKeyMessageAcceptancePolicy {
    fun decide(
        callMatches: Boolean,
        channelMatches: Boolean,
        membershipEpochMatches: Boolean,
        authorizedSender: Boolean,
        messageEpoch: Int,
        currentEpoch: Int,
    ): CallKeyMessageDisposition {
        if (!callMatches || !channelMatches || !membershipEpochMatches) {
            return CallKeyMessageDisposition.DISCARD
        }
        if (messageEpoch > currentEpoch) return CallKeyMessageDisposition.DEFER_FUTURE_EPOCH
        if (messageEpoch < currentEpoch || !authorizedSender) return CallKeyMessageDisposition.DISCARD
        return CallKeyMessageDisposition.PROCESS
    }
}

internal object EncryptedCallKeyCodec {
    private val magic = "PTTC".toByteArray(StandardCharsets.UTF_8)
    private const val VERSION: Byte = 1

    fun encode(message: EncryptedCallKeyMessage): ByteArray {
        val identity = message.participantIdentity.toByteArray(StandardCharsets.UTF_8)
        require(message.membershipEpoch > 0 && message.callEpoch > 0 && identity.size in 16..128)
        require(when (message.kind) {
            EncryptedCallKeyMessageKind.ANNOUNCEMENT -> message.key?.size == 32
            EncryptedCallKeyMessageKind.ACKNOWLEDGEMENT -> message.key?.size == 16
        })
        return ByteBuffer.allocate(63 + identity.size + (message.key?.size ?: 0)).apply {
            put(magic); put(VERSION); put(message.kind.wire)
            putUuid(message.messageId); putUuid(message.channelId); putUuid(message.callId)
            putInt(message.membershipEpoch); putInt(message.callEpoch)
            put(identity.size.toByte()); put(identity); message.key?.let(::put)
        }.array()
    }

    fun decode(bytes: ByteArray, senderAci: String, senderDeviceId: Int): EncryptedCallKeyMessage {
        require(bytes.size >= 63 && bytes.copyOfRange(0, 4).contentEquals(magic) && bytes[4] == VERSION)
        require(senderDeviceId in 1..2)
        UUID.fromString(senderAci)
        val buffer = ByteBuffer.wrap(bytes)
        buffer.position(5)
        val kind = requireNotNull(EncryptedCallKeyMessageKind.fromWire(buffer.get()))
        val messageId = buffer.getUuid()
        val channelId = buffer.getUuid()
        val callId = buffer.getUuid()
        val membershipEpoch = buffer.int
        val callEpoch = buffer.int
        val identitySize = buffer.get().toInt() and 0xff
        require(identitySize in 16..128 && buffer.remaining() >= identitySize)
        val identityBytes = ByteArray(identitySize).also(buffer::get)
        val key = ByteArray(buffer.remaining()).also(buffer::get).takeIf(ByteArray::isNotEmpty)
        val message = EncryptedCallKeyMessage(
            messageId, channelId, membershipEpoch, callId, callEpoch, kind,
            identityBytes.toString(StandardCharsets.UTF_8), key, senderAci.lowercase(), senderDeviceId,
        )
        require(encode(message).contentEquals(bytes))
        return message
    }

    private fun ByteBuffer.putUuid(value: UUID) { putLong(value.mostSignificantBits); putLong(value.leastSignificantBits) }
    private fun ByteBuffer.getUuid() = UUID(long, long)
}

/** Device-encrypted durable inbox format for call coordination consumed across app components. */
internal object EncryptedCallKeyQueueCodec {
    private val magic = "PTTQ".toByteArray(StandardCharsets.UTF_8)
    private const val VERSION: Byte = 1
    const val MAX_MESSAGES = 128

    fun encode(messages: List<EncryptedCallKeyMessage>): ByteArray {
        require(messages.size <= MAX_MESSAGES)
        val rows = messages.map { message ->
            require(message.senderDeviceId in 1..2)
            UUID.fromString(message.senderAci)
            message to EncryptedCallKeyCodec.encode(message)
        }
        return ByteBuffer.allocate(7 + rows.sumOf { 19 + it.second.size }).apply {
            put(magic); put(VERSION); putShort(rows.size.toShort())
            rows.forEach { (message, body) ->
                val sender = UUID.fromString(message.senderAci)
                putLong(sender.mostSignificantBits); putLong(sender.leastSignificantBits)
                put(message.senderDeviceId.toByte()); putShort(body.size.toShort()); put(body)
            }
        }.array()
    }

    fun decode(bytes: ByteArray?): List<EncryptedCallKeyMessage> {
        if (bytes == null || bytes.isEmpty()) return emptyList()
        require(bytes.size >= 7 && bytes.copyOfRange(0, 4).contentEquals(magic) && bytes[4] == VERSION)
        val buffer = ByteBuffer.wrap(bytes).apply { position(5) }
        val count = buffer.short.toInt() and 0xffff
        require(count <= MAX_MESSAGES)
        val messages = buildList(count) {
            repeat(count) {
                require(buffer.remaining() >= 19)
                val sender = UUID(buffer.long, buffer.long).toString().lowercase()
                val deviceId = buffer.get().toInt() and 0xff
                val size = buffer.short.toInt() and 0xffff
                require(deviceId in 1..2 && size in 63..255 && buffer.remaining() >= size)
                val body = ByteArray(size).also(buffer::get)
                add(EncryptedCallKeyCodec.decode(body, sender, deviceId))
            }
        }
        require(!buffer.hasRemaining())
        return messages
    }
}
